presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/OrcStorageManager.java

```diff
@@ -0,0 +1,593 @@
+        private final int bucketNumber;
+        private final List<Long> columnIds;
+        private final List<Type> columnTypes;
+        private final CompressionType compressionType;
+
+        private final List<File> stagingFiles = new ArrayList<>();
+        private final List<ChunkInfo> chunks = new ArrayList<>();
+        private final List<ListenableFuture<?>> futures = new ArrayList<>();
+
+        private boolean committed;
+        private OrcFileWriter writer;
+        private long chunkId;
+
+        public OrcStoragePageSink(
+                long transactionId,
+                long tableId,
+                int bucketNumber,
+                List<Long> columnIds,
+                List<Type> columnTypes,
+                CompressionType compressionType)
+        {
+            this.transactionId = transactionId;
+            this.tableId = tableId;
+            this.bucketNumber = bucketNumber;
+            this.columnIds = ImmutableList.copyOf(requireNonNull(columnIds, "columnIds is null"));
+            this.columnTypes = ImmutableList.copyOf(requireNonNull(columnTypes, "columnTypes is null"));
+            this.compressionType = requireNonNull(compressionType, "compressionType is null");
+        }
+
+        @Override
+        public void appendPages(List<Page> pages)
+        {
+            createWriterIfNecessary();
+            writer.appendPages(pages);
+        }
+
+        @Override
+        public void appendPages(List<Page> pages, int[] pageIndexes, int[] positionIndexes)
+        {
+            createWriterIfNecessary();
+            writer.appendPages(pages, pageIndexes, positionIndexes);
+        }
+
+        @Override
+        public boolean isFull()
+        {
+            if (writer == null) {
+                return false;
+            }
+            return (writer.getRowCount() >= maxChunkRows) || (writer.getUncompressedSize() >= maxChunkSize.toBytes());
+        }
+
+        @Override
+        public void flush()
+        {
+            if (writer != null) {
+                writer.close();
+
+                File file = storageService.getStagingFile(chunkId);
+
+                chunkRecorder.recordCreatedChunk(transactionId, tableId, chunkId, file.length());
+
+                futures.add(chunkStoreManager.submit(chunkId, file));
+
+                long rowCount = writer.getRowCount();
+                long uncompressedSize = writer.getUncompressedSize();
+
+                chunks.add(createChunkInfo(chunkId, bucketNumber, file, rowCount, uncompressedSize));
+
+                writer = null;
+                chunkId = 0;
+            }
+        }
+
+        @Override
+        public ListenableFuture<List<ChunkInfo>> commit()
+        {
+            checkState(!committed, "already committed");
+            committed = true;
+
+            flush();
+
+            return whenAllComplete(futures).call(() -> {
+                for (ChunkInfo chunk : chunks) {
+                    writeChunk(chunk.getChunkId());
```

Comment:
I observe this error on raptor-dev:

com.facebook.presto.spi.PrestoException: Failed to move chunk file
        at com.facebook.presto.raptorx.storage.OrcStorageManager.writeChunk(OrcStorageManager.java:325)
Caused by: java.nio.file.NoSuchFileException: /raptor/staging/635606 -> /raptor/storage/1c/25/635606

I am guessing it's such time series:
transaction1 calls OrcStoragePageSink.commit(), fill futures, which runs async;
Some logic calls OrcStoragePageSink.rollback(),  stagingFile.delete()  done;
futures got finished and then calls writeChunk(), which does Files.move(staging, storage); Then throw exception

If the above is due to intentional killing, like human user kills, or system killing for resource management, it may be fine, since the transaction's data is gone any way. But I am not familiar with all calling paths, need to confirm with you.

And if it's indeed intentional killing, shall we consider output accurate error message, when the transaction is killed, like
 "com.facebook.presto.spi.PrestoException: Failed to move chunk file; 
chunkXXX in transactionYYY,  transaction already aborted by user"
Or "chunkXXX in transactionYYY,  transaction already killed by ResourceManager"

Reply:
I tracked down the log and it seems this is being logged from Guava `AggregateFuture`:

```
2018-09-15T08:45:00.313-0700	ERROR	raptor-commit-2668	com.google.common.util.concurrent.AggregateFuture	Got more than one input Future failure. Logging failures after the first
```

The fact that Guava is logging here is very unexpected. I understand why they do it (the combined future only contains the first failure), but I still think it's the wrong behavior. There are some complaints about this but no resolution yet: https://github.com/google/guava/issues/2134

I'm still digging through the code to figure out how we can work around this.

Reply:
I kind of think the "com.facebook.presto.spi.PrestoException: Failed to move chunk file" is by TableWriterOperator.java line 300:
Collection<Slice> fragments = getFutureValue(finishFuture);

This "finishFuture" has 2 parts:  // from OrcStorageManager line 550
1. whenAllComplete(futures)
2. for (ChunkInfo chunk : chunks) { writeChunk}

TableWriterOperator.java line 300 doesn't log part1, due to whenAllComplete. It only logs part2